src/common/cockpitwebinject.c

```diff
@@ -84,62 +87,68 @@ cockpit_web_inject_push (CockpitWebFilter *filter,
   const gchar *mark, *data, *pos;
   gsize mark_len, data_len, at;
   GBytes *bytes;
+  gsize written;
 
   /* Stop searching once injected */
-  if (self->inject)
+  mark = g_bytes_get_data (self->marker, &mark_len);
+  data = g_bytes_get_data (block, &data_len);
+  written = at = 0;
+
+  while (self->injected < self->maximum && at < data_len)
     {
-      mark = g_bytes_get_data (self->marker, &mark_len);
-      data = g_bytes_get_data (block, &data_len);
-      at = 0;
+      if (self->partial)
+        pos = data + at;
+      else
+        pos = memchr (data + at, mark[0], data_len - at);
 
-      while (at < data_len)
-        {
-          if (self->partial)
-            pos = data + at;
-          else
-            pos = memchr (data + at, mark[0], data_len - at);
+      /* Couldn't find the character anywhere? */
+      if (!pos)
+        break;
 
-          /* Couldn't find the character anywhere? */
-          if (!pos)
+      for (at = (pos - data); self->partial < mark_len && at < data_len; self->partial++, at++)
```

Comment:
This is a property of the code before this pull request. I've added a test for this case in my checkout, and it does indeed fail.